community/rfcs/24-6-17-OPEA-001-Components-Deployment-on-Kubernetes.md

```diff
@@ -0,0 +1,73 @@
+RFC Template
+==================
+
+## RFC Title
+
+Deploy GenAIComps on Kubernetes
+
+## RFC Content
+
+### Author
+
+[yongfengdu](https://github.com/yongfengdu),[leslieluyu](https://github.com/leslieluyu)
+
+### Status
+
+`Under Review`
+
+### Objective
+
+Define how the GenAIComps/Microservices should be deployed on kubernetes.
+
+Non-Goal: Pipeline to connect all microservices is out of scope.
+
+### Motivation
+
+GenAIComps defined building blocks for AI components, and provided instructions on how to run with docker/docker-compose.
+This proposal will provide a way to run these components on kubernetes.
+
+### Design Proposal
+
+Yaml files are the straightforward way to deploy applications on kubernetes,
+but when we are providing deployment for more examples in GenAIExamples,
+there will be a lot of duplicated codes. This make the yaml files hard to
+maintain.
+
+The proposal is to provide helm chart for GenAIComps components, and define a
+few variables like model name, so we can generate the yaml files automatically
+by "helm template" for each GenAIExamples deployment.
+
+In detail, the helm chart for each component will be maintained at this directory:
+
+GenAIInfra/helm-charts/common
+
+For example llm-uservice, then we can generate the yaml files for CodeTrans this way:
+
+```console
+cd GenAIInfra/helm-charts/common
+helm dependency update llm-uservice
+export HFTOKEN="insert-your-huggingface-token-here"
+export MODELDIR="/mnt"
+export MODELNAME="HuggingFaceH4/mistral-7b-grok"
+helm template codetrans llm-uservice --set global.HUGGINGFACEHUB_API_TOKEN=${HFTOKEN} --set tgi.volume=${MODELDIR} --set tgi.LLM_MODEL_ID=${MODELNAME} > ../../manifests/CodeTrans/xeon/llm.yaml
+```
+
+We'll keep a copy of the generated yaml files at GenAIInfra/manifests/${WORKLOAD}/[xeon|gaudi] in case anyone want to use it directly with "kubectl apply -f".
```

Comment:
Thanks @yongfengdu for this proposal.  Is it possible we host only one manifest yaml for each of the microservices?  Then for each sample, we just put the variables into a configmap.  Like GenAIInfra/manifests/${WORKLOAD}/[xeon|gaudi]

Reply:
For CodeTrans, there is only one microservice which is llm, the llm.yaml includes service/deployment for llm and tgi. See this for details (https://github.com/opea-project/GenAIInfra/pull/97/files#diff-51820540a88c4f7d4bbe5e23ba25232ab847632cd52224d247006f3be966b411)
It might be more appropriate to use ChatQnA as example here. In chatqna, we could have four components as microservices - llm, reranking, retrieving, embedding.
@leslieluyu and @zhlsunshine any suggestions on the chatqna case?

Reply:
Hi @yongfengdu, thanks for the alignment, I have some suggestions below:
1. Each micro-service (such as tgi-svc, tei-svc and retrieve-svc, etc) can have only one Yaml template file.
2. Each example (such as ChatQnA, CodeGen and DocSum, etc) can have configmaps to hold different parameters for micro-service template files.

These suggestions mean that we have one set of Yaml template files for micro-services and each example should need different configmaps to support different parameters for micro-services. Does it make sense?

Reply:
Yes, this makes sense.
I was implementing this structure with helm chart. And the design is to provide a different values.yaml file for different workload (Chatqna, CodeGen).
Is that possible for GMC to try with helm chart way of deploying the microservice component? Or one alternative is to for GMC to generate manifests with command "helm template", this way we can remove the duplicated yaml files in each workload directory.
I wasn't sure if that's doable to use configmap, we used config map to pass Environment variables to yaml files previously, but if we're defining more variables (Resource limit, docker images for Gaudi etc), I'm afraid we're expanding too much and it grows into another helm like template.

Reply:
I just updated the RFC and added the flow graph to show how to export manifests from helm-chart.
please check @zhlsunshine  @irisdingbj